reports/module_cost_by_type.go

```diff
@@ -0,0 +1,288 @@
+//   Copyright 2019 MSolution.IO
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+
+package reports
+
+import (
+	"context"
+	"database/sql"
+	"encoding/json"
+	"fmt"
+	"github.com/360EntSecGroup-Skylar/excelize"
+	"github.com/olivere/elastic"
+	"github.com/trackit/jsonlog"
+	"github.com/trackit/trackit/aws"
+	"github.com/trackit/trackit/aws/s3"
+	"github.com/trackit/trackit/aws/usageReports/history"
+	"github.com/trackit/trackit/es"
+	"github.com/trackit/trackit/users"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+)
+
+const costByTypeSheetName = "Cost by type"
+const maxAggregationSize = 0x7FFFFFFF
+
+var costByTypeModule = module{
+	Name:          "Cost by type",
+	SheetName:     costByTypeSheetName,
+	ErrorName:     "costByTypeError",
+	GenerateSheet: generateCostByTypeSheet,
+}
+
+type QueryParams struct {
+	AccountList []string  `json:"awsAccounts"`
+	IndexList   []string  `json:"indexes"`
+	DateBegin   time.Time `json:"begin"`
+	DateEnd     time.Time `json:"end"`
+}
+
+type esCostByTypesResult struct {
+	Buckets []struct {
+		Key      string `json:"key"`
+		Services struct {
+			Buckets []struct {
+				Key       string `json:"key"`
+				Resources struct {
+					Buckets []struct {
+						Key      string `json:"key"`
+						DocCount int    `json:"doc_count"`
+						Cost     struct {
+							Value float32 `json:"value"`
+						} `json:"cost"`
+					} `json:"buckets"`
+				} `json:"resources"`
+			} `json:"buckets"`
+		} `json:"services"`
+	} `json:"buckets"`
+}
+
+func generateCostByTypeSheet(ctx context.Context, aas []aws.AwsAccount, date time.Time, tx *sql.Tx, file *excelize.File) (err error) {
+	if date.IsZero() {
+		date, _ = history.GetHistoryDate()
+	}
+	return costByTypeGenerateSheet(ctx, aas, date, file, tx)
+}
+
+func costByTypeGenerateSheet(ctx context.Context, aas []aws.AwsAccount, date time.Time, file *excelize.File, tx *sql.Tx) (err error) {
+	data, err := costByTypeGetData(ctx, aas, date, tx)
+
+	if err == nil {
+		costByTypeInsertDataInSheet(file, costByTypeSheetName, data)
+		return nil
```

Comment:
You can remove this line, if `err` is `nil`, you will return `nil` with the next line